### ISSUE TITLE
Provide access to parser

### ISSUE THREAD

**Issue by yvdh** [dannyhaak/TagDataTranslation#3]
Hello,

first of all: great work!

I'm usually using TranslateDetails, but most often than not I am not interested in the actual translation but rather the 'parsed' information of the input string. This incurs some unnecessary overhead, besides requiring extra parameters (outputformat)

Looking at the code it seems this could be achieved by making the ProcessInput  method of the TDTEngine public.

Cheers
Yves

**Comment by dannyhaak** [comment]
Hi,

Thanks! Out of curiosity, what are you using the package for?

I have made the change you requested, and updated both the repository and the NuGet package (version 1.1). If you have any questions, let me know.

Regards,
Danny

**Comment by yvdh** [comment]
Hallo Danny,

thanks for the quick addition, I will check it out later this week.

We are setting up traceability in the hospital with an EPCIS framework and an BPM-engine. EPCIS only stores pure URI's, and I have setup a GS1 Masterdata repository en web API (we have currently chosen not to store this in the CBV of EPCIS because we are on EPCIS 1.1). I am using your package to provide general TDT services in that web API,  mainly from barcode <=> Pure URI's (and be future proof with regards to RFID). Note that internally I am not using the TDT services but a more pragmatic and fast translation based on the prior knowledge of our prefix, serials, etc. ... We will evaluate this if necessary!

Cheers

Yves

Dr. Sc. Yves Vander Haeghen
Dienst ICT, Universitair Ziekenhuis Gent,
De Pintelaan 185,
9000 Gent.
Tel. ++ 32 9 332 62 78
Route 5017

**Comment by dannyhaak** [comment]
Hi Yves,

Ah, that's interesting! So you are using EPCIS in a non-RFID context? Maybe
good for you to know, I am currently co-chairing the EPCIS 2.0 working
group, in which we are planning a JSON(-LD) and REST implementation of
EPCIS, which will make it much more developer-friendly. If you are
interested, I'd be happy to share more about this.

Regards,
Danny

**Comment by yvdh** [comment]
Hello Danny,

Yes we will be using mainly barcodes, but RFID is not out of scope completely. It's more a 'learn to walk before we run' kind of strategy.
I am sure there are use-cases were RFID will be the only viable way to tag, like some type of medical instruments, with high volumes and when the physical paths of items are restricted (in and outgoing doors) like in the central warehouse.

Anyway, adding JSON REST seems a great idea, we have developped several applications which are all REST-based, and they seem to work very well. My GS1 Masterdata is purely rest based, with no database access by third-party programs. Our label print functionality in the hospital is also JSON REST based, and completely driverless. JSON is just simpler than SOAP to implement and configure ...

Concerning GS1 Masterdata, this seems to me to be an underestimated issue that is not completely covered by the CBV documention and I am wandering how people are tackling this in general (in combination with EPCIS) ...

Long story short, I am very much interested in what you and other people are doing with reagrds to EPCIS, and I am more than willing to share my experiences when I can! I very much believe GS1 and EPCIS (with GDSN) are the way forward for track and trace and better safety and quality in a hospital. I have been promoting this for years in the hospital, but it's a slow proces ...

Cheers,


Dr. Sc. Yves Vander Haeghen
Dienst ICT, Universitair Ziekenhuis Gent,
De Pintelaan 185,
9000 Gent.
Tel. ++ 32 9 332 62 78
Route 5017

**Comment by dannyhaak** [comment]
Hi Yves,

That sounds like the best strategy. A lot of people do it the other way
around, and that is not a recipe for succes.

What do you mean with 'MasterData'? The bizStep etc. in the CBV is not
considered master data in the traditional way. There are capabilities for
MasterData in EPCIS, but they are currently under review for 2.0, as it
doesn't really feel right to manage master data in EPCIS events.

I think you would be very much welcomed to join the EPCIS working group,
which actually lacks people with actual implementation experience from an
end-user perspective. We have a call every week or two of an hour, and you
are free to only join on the topics that are of interest for you. As we are
drafting the REST/JSON spec, feedback is especially welcome on
implementation decisions. Furthermore, there will be work ongoing to extend
the CBV with more BizStep, Disposition, etc.; and there is a work stream to
work on including sensor data in EPCIS. See attached leaflet for more
information. I think you need to be a GS1 member organisation to join - but
not sure. Can check for you.

Regards,
Danny

**Comment by yvdh** [comment]
Hello,

My hospital is a member organization (we have a prefix key), and I am very much interested in participating in discussions regarding EPCIS and GS1, and meeting fellow implementers!

To answer your question, with 'MasterData' I mean the metadata associated with Pure URI's as stored in EPCIS (SGLN, GTIN, SGTIN, GSRNP, SSCC, GRAI)

E.g. for urn:epc:id:sgln:5400531.11986.0 (barcode = 41454005311198612540)  I would receive from the web service:

{
    "Building": "P1-P2",
    "Floor": "0",
    "Description": "Functieonderzoek",
    "Uri": "urn:epc:id:sgln:5400531.11986.0",
    "Id": 23086,
    "Name": "004",
    "ExternalId": {
      "Id": "137743",
      "Source": "FMIS"
    }
  }

These are to be used in user interfaces, reports, statistics, ...

I tend to agree they have no place in EPCIS, but I also think that providing a standard method of providing such 'metadata' services could be useful.

Currently I have something akin to the next swagger screenshot. There is also a TDT-service controller that uses your code to translate between different representations of GS1 keys and EPC tags.

[cid:image001.png@01D4C54E.E23C7180]

[cid:image002.png@01D4C54F.21121B30]

Cheers

Yves

Dr. Sc. Yves Vander Haeghen
Dienst ICT, Universitair Ziekenhuis Gent,
De Pintelaan 185,
9000 Gent.
Tel. ++ 32 9 332 62 78
Route 5017

**Comment by dannyhaak** [comment]
Hi Yves,

That sounds good. Can you drop me an e-mail via regular mail at
mail@dannyhaak.nl; I'll make sure you get introduced to the group.

I *think* that your master data can be expressed as EPCIS 1.2 compatible
master data, and stored and queried via MasterData documents. Not all
fields are standardised, but like EPCIS Events, Master Data can be extended
with additional fields. (And, yes, the XML representation is a bit
cumbersome, but JSON support will solve that). You can then have it stored
outside of the EPCIS events, and still query it for usage in the UI. I
think that is a very valid use case.

Do you use a commercial EPCIS repository, or did you build your own one?

(The images didn't end up in the GitHub issue - so if you can include them
again in a direct e-mail, that would be great).

Thanks,
Danny

**Comment by dannyhaak** [comment]
Yves,

I didn't hear back from you. Just checking whether your e-mail wasn't lost in the cloud?

Regards,
Danny
